basis/basis.py

```diff
@@ -0,0 +1,172 @@
+    its position in the list points.
+    ``levels`` property gives the Smolyak indexes for each
+    grid level.
+
+    Parameters
+    ----------
+    max_exactness : int
+        level of exactness class will describe
+
+    """
+
+    def __init__(self,max_exactness):
+        self._max_exactness = max_exactness
+        self._points = []
+        self._levels = []
+        self._update()
+
+    @property
+    def max_exactness(self):
+        """Maximum level of exactness described"""
+        return self._max_exactness
+
+    @property
+    def points(self):
+        """Extrema assigned to Smolyak indices"""
+        return self._points
+
+    @property
+    def levels(self):
+        """index of each point that belongs to each grid level"""
+        return self._levels
+
+    @max_exactness.setter
+    def max_exactness(self, max_exactness):
+        """Set maximum level of exactness
+        Sets maximmum level of exactness class instance will describe, will
+        extend or truncate properties points and levels to match new value
+
+        Parameters
+        ----------
+        max_exactness : int
+            new level of exactness class should describe
+        """
+        if max_exactness > self._max_exactness:
+            # extend
+            self._max_exactness = max_exactness
+            # update other properties
+            self._update()
+        elif max_exactness < self._max_exactness:
+            # truncate
+            self._max_exactness = max_exactness
+            # update other properties
+            points_keep = sum(
+                    [len(x) for x in self._levels[:self._max_exactness+1]])
+            self._points = self._points[:points_keep]
+            self._levels = self._levels[:self._max_exactness+1]
+
+
+    @abc.abstractmethod
+    def _update(self):
+        """Update properties described by level of exactness
+        Compute points of basis function and the Smolyak index of
+        each grid level between 0 and max_exactness + 1 
+        """
+        pass
+
+    @abc.abstractmethod
+    def basis(self,x,n):
+        """Compute term of basis function
+        Returns the output of an nth degree basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        n : int
+            degree
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis Function of Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a set of polynomials described by the
+    following recurrence relation:
+
+    ..math:
+        T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2xT_n(x) - T_{n-1}(x)
+
+    The extrema of the polynomials are calculated via the following
+    equation:
+
+    ..math:
+        x_{n,j}^* = -\cos\left((\frac{j-1}{n-1}\pi\right), j = 1,...,n
+    """
+
```

Comment:
Here you can add a constructor with an argument named like `max_n` indicating the maximum value of `n` allowed for the basis function (your previous `max_exactness`).

Reply:
max_n isn't same as max_exactness. 
Would it be sufficient to add an equation specifying how `max_n` and `max_exactness` are related? 
`max_n = 2**(max_exactness-1)+1`

Reply:
Maybe `num_levels`? Since based on what value you choose here, there will be a certain number of levels of points, right?

Reply:
Yes. `max_exactness` is supposed to be the number of grid levels described in `self._levels` minus 1. That's its relationship with the number of grid levels. I called it `max_exactness` because it's meant to be exactly equal to the exactness parameter in `IndexGrid`, and I feel it's important if that relationship is very clear in the variable names.

Do you want `max_exactness` to be renamed `num_levels` and they have the same value? Should `num_levels` be `max_exactness` + 1?

Reply:
Let’s discuss together at our next meeting. We planned to generalize the IndexGrid when you weren’t there, and so not all IndexGrids will have an exactness variable.